### PR TITLE
Add sync_metadata_to_node UDF

### LINKED ISSUE #1011: Add a UDF to copy the current metadata from master to worker
For MX, we need to copy metadata from master to worker and mark that worker to receive future DDL changes. 

Basically, after the cluster is set, the user should be able to call a function which marks a worker node as containing the metadata and copies the current metadata snapshopt to that worker.

Note that we need to copy the metadata related to only mx tables. 

Here are @marcocitus's notes about the issue:

> A change I'd like to propose compared to citus-mx is to rename the nodeactive column to hasmetadata and rename cluster_activate_node(nodename,nodeport) to something like sync_metadata_to_node(nodename, nodeport), with the intent of decoupling the concepts of metadata being synced to a node and nodes being active, which are coupled in MX.
> 
> If we couple metadata syncing with the concept of a node being active (present in pg_dist_node), then syncing all pg_dist_node changes to all nodes becomes a requirement. However, in a non-MX Citus cluster (all tables have repmodel 'c'), there is no reason to sync changes to pg_dist_node or other metadata tables and doing so could create an operational burden. For example, you would not be able to add a new node if any other node is down, since syncing would fail. This burden is acceptable for MX clusters, which assume nodes are eventually replaced by their secondary, but would make regular Citus clusters much harder to operate.
> 
> The main functionality of cluster_activate_node(nodename,nodeport) is actually to a) sync metadata to the specified node, b) ensure all future metadata changes are propagated to the specified node. Therefore, renaming it to sync_metadata_to_node seems reasonable.
> 
> Apart from the renames, only a few small changes to the existing code are required, mainly:
> 
> In worker_transaction.c we should send only to nodes with the hasmetadata flag enabled, it might make sense to rename the file and functions to reflect this change.
> master_add_node should propagate the insert into pg_dist_node to all nodes with syncmetadata enabled. We currently send this insert in cluster_activate_node, but we might put shards on the node without calling sync_metadata_to_node, so we need to do this in master_add_node instead.
> sync_metadata_to_node should set the hasmetadata flag to true (instead of setting nodeactive), but propagating this update to workers is optional and can be skipped for now, since no operations on workers use this flag.
> An additional benefit of decoupling metadata syncing is that we could easily add the explicit concepts of shards-only (data) nodes or distributed tables-only (aggregator) nodes if we'd wish to do so in the future.
> 
> Some things to note:
> 
> - In 6.0 we added support for foreign keys. When we sync metadata to workers then we need to make sure that the tables are created before the foreign keys are added. Otherwise foreign key creation would error out.
> - In 6.0 we added support for truncate through triggers. We need to explicitly add these triggers as part of the DDL commands that we send to the workers.
> - We ran into an issue in MX with the schema being created by super user, which made the schema inaccessible to regular users. We should make sure that the CREATE SCHEMA command specifies an AUTHORIZATION: https://www.postgresql.org/docs/current/static/sql-createschema.html

### PR DESCRIPTION
**Pull request by aamederen**
This change adds sync_metada_to_node UDF which copies the metadata about nodes and MX tables  from master to the specified worker, sets its local group ID and marks its hasmetadata to true to allow it receive future DDL changes.

Fixes #1011 

Remaining Work:
- [x] Remove `hasmetadata` propagation
- [x] UDF for marking `hasmetadata=false` of a worker.
- [x] Prevent transactions
- [x] Propagate truncate triggers
- [x] Correct foreign key constraints propagation
- [x] make sure that the CREATE SCHEMA command specifies an AUTHORIZATION
- [x] Self-review and final touches
- [x] Rename UDFs

### CODE REVIEW

File: src/test/regress/sql/multi_metadata_snapshot.sql [hunk 1]
```diff
@@ -59,4 +59,51 @@ SELECT unnest(master_metadata_snapshot());
...
+
+-- Run sync_metadata_to_node and check that it marked hasmetadata for that worker
+SELECT sync_metadata_to_node('localhost', :worker_1_port);
+SELECT nodeid, hasmetadata FROM pg_dist_node WHERE nodename='localhost' AND nodeport=:worker_1_port;
+
+-- Check that the metadata has been copied to the worker
+\c - - - :worker_1_port
+SELECT * FROM pg_dist_local_group;
+SELECT * FROM pg_dist_node ORDER BY nodeid;
+SELECT * FROM pg_dist_partition ORDER BY logicalrelid;
```
Comment (marcocitus) [inline src/test/regress/sql/multi_metadata_snapshot.sql]: Should probably also do a \d on the distributed tables here.

File: src/backend/distributed/worker/worker_truncate_trigger_protocol.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+worker_create_truncate_trigger(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+
+	EnsureSuperUser();
+
+	/* Ensure that this node is a worker with metadata */
+	if (GetLocalGroupId() == 0)
+	{
+		ereport(ERROR, (errmsg("this node is not a metadata worker")));
```
Comment (marcocitus) [inline src/backend/distributed/worker/worker_truncate_trigger_protocol.c]: We'll probably want this logic in a lot of places, so maybe create an EnsureCoordinator() function.

Comment (marcocitus) [inline src/backend/distributed/worker/worker_truncate_trigger_protocol.c]: No wait, we want the opposite in a lot of places :) never mind!

File: src/backend/distributed/worker/worker_truncate_trigger_protocol.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	/* Ensure that the specified table is an MX table */
+	if (!ShouldSyncTableMetadata(relationId))
```
Comment (marcocitus) [inline src/backend/distributed/worker/worker_truncate_trigger_protocol.c]: I think we could actually reuse this UDF for the upgrade script, so maybe not put too many checks on it.

File: src/test/regress/sql/multi_metadata_snapshot.sql (hunk 1 above)
Comment (aamederen) [inline src/test/regress/sql/multi_metadata_snapshot.sql]: I added `\d tablename` here. Is there a way to limi `\d` to distributed tables only?

File: src/backend/distributed/worker/worker_truncate_trigger_protocol.c (hunk 3 above)
Comment (marcocitus) [inline src/backend/distributed/worker/worker_truncate_trigger_protocol.c]: Upgrading from Citus <=5.2 to >=6.0

File: src/test/regress/sql/multi_metadata_snapshot.sql (hunk 1 above)
Comment (marcocitus) [inline src/test/regress/sql/multi_metadata_snapshot.sql]: No, it's implemented by psql, so not much we can do about it.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 4]
```diff
@@ -27,14 +31,91 @@
...
+
+	PreventTransactionChain(true, "sync_metadata_to_node");
+
+	workerNode = FindWorkerNode(nodeNameString, nodePort);
+
+	if (workerNode == NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("you cannot sync metadata to a non-existent node"),
+						errhint("First, add the node with master_add_node().")));
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: Could give the exact command here, e.g. `... with: SELECT master_add_node(%s,%d)", nodeNameString, nodePort)`

Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: Added

File: src/backend/distributed/worker/worker_truncate_trigger_protocol.c (hunk 3 above)
Comment (aamederen) [inline src/backend/distributed/worker/worker_truncate_trigger_protocol.c]: Removed the checks.

File: src/backend/distributed/master/master_metadata_utility.c [hunk 5]
```diff
@@ -765,6 +767,50 @@ TableOwner(Oid relationId)
 
 
 /*
+ * SchemaOwner returns the name of the owner of the specified schema.
+ */
+char *
+SchemaOwner(Oid schemaId)
```
Comment (marcocitus) [inline src/backend/distributed/master/master_metadata_utility.c]: Seems that we don't have a nice file for this function right now. Maybe just keep it in master_node_protocol where it is called and make it static.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 6]
```diff
@@ -27,14 +31,123 @@
...
+stop_sync_metadata_to_node(PG_FUNCTION_ARGS)
+{
+	text *nodeName = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	char *nodeNameString = text_to_cstring(nodeName);
+	WorkerNode *workerNode = NULL;
+
+	EnsureSuperUser();
+
+	PreventTransactionChain(true, "stop_sync_metadata_to_node");
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: This doesn't seem necessary. MarkNodeHasMetadata can safely be called from a transaction.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 7]
```diff
@@ -395,24 +550,91 @@ NodeDeleteCommand(uint32 nodeId)
...
+ */
+static char *
+TruncateTriggerCreateCommand(Oid relationId)
+{
+	StringInfo triggerCreateCommand = makeStringInfo();
+	char *tableName = generate_qualified_relation_name(relationId);
+
+	appendStringInfo(triggerCreateCommand,
+					 "SELECT worker_create_truncate_trigger('%s')",
+					 tableName);
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: needs escaping (using quote_qualified_identifier)

Comment (aamederen) [inline src/backend/distributed/metadata/metadata_sync.c]: we already call `generate_qualified_relation_name`, which inside calls `quote_qualitief_identifier`, am I wrong?

File: src/backend/distributed/master/master_metadata_utility.c (hunk 5 above)
Comment (aamederen) [inline src/backend/distributed/master/master_metadata_utility.c]: It seemed like a utility function but making it static until someone needs that makes sense.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 7 above)
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c]: Oops, I meant quote_literal_cstr. These are two different quotes. Also, the template should not contain quotes:

```
"SELECT worker_create_truncate_trigger(%s)", quote_literal_cstr(tableName))"
```

**Pull request merged**